File: 0061-rotate-list/0061-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(k == 0 || head == null || head.next == null) return head;
        int len = 1;
        ListNode end = head, prev = head;
        while(end.next != null) {
            end = end.next;
            ++len;
        }
        if(k % len == 0) return head;
        if(k < len) k = len - k;
        else k = len - (k % len);
        end.next = head;
        while(k-- > 0) {
            prev = head;
            head = head.next;
        }
        prev.next = null;
        return head;
    }
}

File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == NULL or head->next == NULL) return head;
        int length = 1;
        ListNode *current = head;
        while(current->next != NULL) {
            ++length;
            current = current->next;
        }
        if(k == length) return head;
        if(k > length) k = k % length;
        k = length - k;
        int i = 0;
        current->next = head;
        int cnt = k - 1;
        current = head;
        while(cnt--) 
            current = current->next;
        head = current->next;
        current->next = NULL;
        return head;
    }
};